src/geo_interface.jl

```diff
@@ -39,21 +43,47 @@ function GeoInterface.getgeom(
     getGeometry(geom, i)
 end
 GeoInterface.getgeom(::MultiPointTrait, geom::MultiPoint, i) = getGeometry(geom, i)::Point
+GeoInterface.getgeom(::MultiPointTrait, geom::MultiPoint) = getGeometry(geom, i)::Point
 GeoInterface.getgeom(::MultiLineStringTrait, geom::MultiLineString, i) =
     getGeometry(geom, i)::LineString
 GeoInterface.getgeom(::MultiPolygonTrait, geom::MultiPolygon, i) =
     getGeometry(geom, i)::Polygon
-GeoInterface.getgeom(
-    ::Union{LineStringTrait,LinearRingTrait},
-    geom::Union{LineString,LinearRing},
-    i,
-) = getPoint(geom, i)
+function GeoInterface.getgeom(::AbstractGeometryTrait, geom::Union{LineString,LinearRing}, i)
+    refs = Ref{Float64}(), Ref{Float64}(), Ref{Float64}() # 3 Refs is faster than a Vector
+    seq = getCoordSeq(geom::Union{LineString, LinearRing})
+    return _get_tuple_point(geom, seq, refs, i)
+end
+function GeoInterface.getgeom(::AbstractGeometryTrait, geom::Union{LineString,LinearRing})
+    context = get_global_context()
+    seq = getCoordSeq(geom, context)
+    n = getSize(seq, context) # Faster thatn `GI.ngeom(geom)` when we already have `seq`
+    # Preallocate refse
+    refs = Ref{Float64}(), Ref{Float64}(), Ref{Float64}()
+    # Pretetermin if there is a z coordinate
+    hasz = hasZ(geom, context)
+    # We specify Unit32 in the generator to avoid unnecesary conversion later
+    return (_get_tuple_point(geom, seq, refs, i, context, hasz) for i in UInt32(1):UInt32(n))
+end
+
+function _get_tuple_point(geom, seq, (x, y, z), i, context = get_global_context(), hasz=hasZ(geom, context))
+    if hasz
+        GC.@preserve x y z begin
```

Comment:
Why is preserve required? You deconstruct that tuple in this method.

Reply:
I dont think so honestly